### PR TITLE
Support localizing data URIs in addition to bucket paths

### PR DESCRIPTION
**Pull request by rtitle**
Issue: https://github.com/DataBiosphere/leonardo/issues/307

Based on: https://github.com/DataBiosphere/leonardo/pull/309

`jupyter_localize_extension` now supports GCS paths or [data URIs
](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs). For example:
```
POST https://leo/notebooks/proj/cluster/api/localize?sync=true
{
  "~/test.txt" : "data:text/plain;base64,SGVsbG8sIFdvcmxkIQ==",
}
```
will localize a file `~/test.txt` containing the string `Hello, World!`

@calbach @zarsky-broad FYI

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: jupyter-docker/jupyter_localize_extension.py [hunk 1]
```diff
@@ -6,34 +6,76 @@
 from tornado.web import HTTPError
 from notebook.base.handlers import IPythonHandler
 from notebook.utils import url_path_join
+from datauri import DataURI
```
Comment (rtitle) [inline jupyter-docker/jupyter_localize_extension.py]: Using this [library](https://pypi.python.org/pypi/python-datauri/) to parse data URIs

File: jupyter-docker/jupyter_localize_extension.py [hunk 2]
```diff
@@ -60,11 +102,12 @@ def post(self):
...
       #run localize synchronous to the HTTP request
       #run_sync() doesn't take arguments, so we must wrap the call in a lambda.
-      success = tornado.ioloop.IOLoop().run_sync(lambda: self.localize(pathdict))
+      failures = tornado.ioloop.IOLoop().run_sync(lambda: self.localize(pathdict))
 
       #complete the request only after localize completes
-      if not success:
-        raise HTTPError(500, "Error occurred during localization. See localization.log for details.")
+      if failures:
+        raise HTTPError(500, "Error occurred localizing the following {} entries: {}. See localization.log for details.".format(
```
Comment (rtitle) [inline jupyter-docker/jupyter_localize_extension.py]: Changed the error message to return a list of the failed localization entries

**Comment by rtitle** [comment]
Passing test run: https://fc-jenkins.dsp-techops.broadinstitute.org/job/leonardo-fiab-test-runner/1779/

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala [hunk 3]
```diff
@@ -393,19 +397,28 @@ trait LeonardoTestUtils extends WebBrowserSpec with Matchers with Eventually wit
...
 
         // create a notebook file to delocalize
         withNewNotebook(cluster) { notebookPage =>
           notebookPage.executeCell(s"""! echo -n "$delocalizeFileContents" > $delocalizeFileName""")
 
           val localizeRequest = Map(
             localizeFileName -> GcsPath(bucketName, bucketObjectToLocalize).toUri,
-            GcsPath(bucketName, GcsObjectName(delocalizeFileName)).toUri -> delocalizeFileName
+            GcsPath(bucketName, GcsObjectName(delocalizeFileName)).toUri -> delocalizeFileName,
+            dataFileName -> s"data:text/plain;base64,${Base64.getEncoder.encodeToString(dataFileContents.getBytes(StandardCharsets.UTF_8))}"
```
Comment (kyuksel) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala]: Hopefully nobody tries to localize something with `¬` 🙏

File: jupyter-docker/jupyter_localize_extension.py [hunk 4]
```diff
@@ -6,34 +6,76 @@
...
+    else:
+      expanded = os.path.expanduser(pathstr)
       try:
         os.makedirs(os.path.dirname(expanded))
       except OSError: #thrown if dirs already exist
         pass
-    return pipes.quote(expanded)
+      return expanded
+
+  def _handle_gcs_uri(self, locout, source, dest):
```
Comment (kyuksel) [inline jupyter-docker/jupyter_localize_extension.py]: Can we be more specific about what `handle` means?

File: src/main/resources/swagger/api-docs.yaml [hunk 5]
```diff
@@ -364,18 +364,26 @@ paths:
           default: false
         - in: body
           description: |
-            JSON object. Keys represent destinations, values represent sources.
-            Paths will be quoted, intermediate local directories made, and then sent to [gsutil cp](https://cloud.google.com/storage/docs/gsutil/commands/cp).
+            JSON object. Keys represent destinations, values represent sources. The following rules apply:
+
+            * If either the source or destination is a GCS path, it will be sent to [gsutil cp](https://cloud.google.com/storage/docs/gsutil/commands/cp). Therefore
+              this can be used to localize a file _from_ a bucket _to_ the notebook server; or delocalize a file _from_ the notebook server _to_ a bucket.
+            * If the source is a Data URI, then the destination file will be a created with the decoded data URI contents. This mode can only be used to localize
```
Comment (kyuksel) [inline src/main/resources/swagger/api-docs.yaml]: Nit: remove _a_ in "will be _a_ created"

File: src/main/resources/swagger/api-docs.yaml [hunk 6]
```diff
@@ -364,18 +364,26 @@ paths:
...
+              file to the notebook server.
```
Comment (kyuksel) [inline src/main/resources/swagger/api-docs.yaml]: Nit: file**s**

File: src/main/resources/swagger/api-docs.yaml (hunk 5 above)
Comment (rtitle) [reply src/main/resources/swagger/api-docs.yaml]: good catch

File: jupyter-docker/jupyter_localize_extension.py (hunk 4 above)
Comment (rtitle) [reply jupyter-docker/jupyter_localize_extension.py]: Maybe `_localize_gcs_uri` would be clearer?

**Pull request closed without merging**